src/cata_imgui.cpp

```diff
@@ -0,0 +1,466 @@
+    // Blue
+    result.Blue = b;
+    // Green
+    result.Green = g;
+    // Red
+    result.Red = r;
+    return result;
+}
+#else
+#include "sdl_utils.h"
+#include <imgui/imgui_impl_sdl2.h>
+#include <imgui/imgui_impl_sdlrenderer2.h>
+
+SDL_Renderer *cataimgui::client::sdl_renderer = nullptr;
+SDL_Window *cataimgui::client::sdl_window = nullptr;
+
+cataimgui::client::client()
+{
+    IMGUI_CHECKVERSION();
+    ImGui::CreateContext();
+    ImGuiIO &io = ImGui::GetIO();
+    ( void )io;
+    io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;     // Enable Keyboard Controls
+    io.ConfigFlags |= ImGuiConfigFlags_NavEnableGamepad;      // Enable Gamepad Controls
+
+    // Setup Dear ImGui style
+    ImGui::StyleColorsDark();
+    ImGui_ImplSDL2_InitForSDLRenderer( sdl_window, sdl_renderer );
+    ImGui_ImplSDLRenderer2_Init( sdl_renderer );
+}
+
+cataimgui::client::~client()
+{
+    ImGui_ImplSDL2_Shutdown();
+}
+
+void cataimgui::client::new_frame()
+{
+    ImGui_ImplSDLRenderer2_NewFrame();
+    ImGui_ImplSDL2_NewFrame();
+
+    ImGui::NewFrame();
+}
+
+void cataimgui::client::end_frame()
+{
+    ImGui::Render();
+    ImGui_ImplSDLRenderer2_RenderDrawData( ImGui::GetDrawData() );
+}
+
+void cataimgui::client::process_input( void *input )
+{
+    ImGui_ImplSDL2_ProcessEvent( static_cast<const SDL_Event *>( input ) );
+}
+
+#endif
+
+void cataimgui::window::draw_colored_text( std::string const &text, const nc_color &color,
+        text_align alignment, float max_width, bool *is_selected, bool *is_focused, bool *is_hovered )
+{
+    nc_color color_cpy = color;
+    draw_colored_text( text, color_cpy, alignment, max_width, is_selected, is_focused, is_hovered );
+}
+
+void cataimgui::window::draw_colored_text( std::string const &text, nc_color &color,
+        text_align alignment, float max_width, bool *is_selected, bool *is_focused, bool *is_hovered )
+{
+    ImGui::PushID( text.c_str() );
+    ImGuiID itemId = GImGui->CurrentWindow->IDStack.back();
+    const auto color_segments = split_by_color( text );
+    std::stack<nc_color> color_stack;
+    color_stack.push( color );
+    size_t chars_per_line = size_t( max_width );
+    if( chars_per_line == 0 ) {
+        chars_per_line = SIZE_MAX;
+    }
+    float cursor_start_x = ImGui::GetCursorPosX();
+#if defined(WIN32) || defined(TILES)
+    size_t char_width = size_t( ImGui::CalcTextSize( " " ).x );
+    chars_per_line /= char_width;
+#endif
+    if( alignment != text_align::Left ) {
+        std::string str_raw = remove_color_tags( text );
+        int fullWidth = ImGui::GetContentRegionAvail().x;
+        float textWidth = ImGui::CalcTextSize( str_raw.c_str() ).x;
+        if( alignment == text_align::Right ) {
+            ImGui::SetCursorPosX( ImGui::GetCursorPosX() + fullWidth - textWidth - 2 );
+        } else if( alignment == text_align::Center ) {
+            ImGui::SetCursorPosX( ImGui::GetCursorPosX() + ( float( fullWidth ) / 2 ) - ( textWidth / 2 ) );
+        }
```

Comment:
It seems `text_align::Left` wraps the text while `text_align::Right` and `text_align::Center` do not always do so. I think it is better to split this into separate alignment and wrapping arguments. If some of the combinations are not used and you don't want to implement them right now you can add some `cata_assert` and documentation, but splitting the flags makes it clearer.

Reply:
the intended behavior is for it to always wrap, i just have not had a use case where right-aligned and center-aligned text needed to be wrapped. maybe we can add a TODO?

Reply:
I just went ahead and deleted support for center-aligned and right-aligned text for now, since it's not currently used anyway